plaso/parsers/cups_ipp.py

```diff
@@ -214,123 +139,162 @@ def _GetStringValue(self, data_dict, name, default_value=None):
 
     return ', '.join(values)
 
-  def _ReadPair(self, parser_mediator, file_object):
-    """Reads an attribute name and value pair from a CUPS IPP event.
+  def _ParseAttribute(self, file_object):
+    """Parses a CUPS IPP attribute from a file-like object.
 
     Args:
-      parser_mediator (ParserMediator): mediates interactions between parsers
-          and other components, such as storage and dfvfs.
       file_object (dfvfs.FileIO): file-like object.
 
     Returns:
-      tuple: contains:
+      tuple[str, object]: attribute name and value.
 
-        str: name or None.
-        str: value or None.
+    Raises:
+      ParseError: if the attribute cannot be parsed.
     """
-    # Pair = Type ID + Name + Value.
-    try:
-      # Can be:
-      #   Group ID + IDtag = Group ID (1byte) + Tag ID (1byte) + '0x00'.
-      #   IDtag = Tag ID (1byte) + '0x00'.
-      type_id = self.INTEGER_8.parse_stream(file_object)
-      if type_id == self.GROUP_END:
-        return None, None
-
-      elif type_id in self.GROUP_LIST:
-        # If it is a group ID we must read the next byte that contains
-        # the first TagID.
-        type_id = self.INTEGER_8.parse_stream(file_object)
-
-      # 0x00 separator character.
-      self.INTEGER_8.parse_stream(file_object)
-
-    except (IOError, construct.FieldError) as exception:
-      parser_mediator.ProduceExtractionError(
-          'unable to parse pair identifier with error: {0!s}'.format(
-              exception))
-      return None, None
+    file_offset = file_object.tell()
+    attribute_map = self._GetDataTypeMap('cups_ipp_attribute')
 
-    # Name = Length name + name + 0x00
     try:
-      pair_name = self.PAIR_NAME.parse_stream(file_object)
-    except (IOError, construct.FieldError) as exception:
-      parser_mediator.ProduceExtractionError(
-          'unable to parse pair name with error: {0!s}'.format(exception))
-      return None, None
+      attribute, _ = self._ReadStructureFromFileObject(
+          file_object, file_offset, attribute_map)
+    except (ValueError, errors.ParseError) as exception:
+      raise errors.ParseError(
+          'Unable to parse attribute with error: {0!s}'.format(exception))
 
-    try:
-      name = pair_name.text.decode('utf-8')
-    except UnicodeDecodeError as exception:
-      parser_mediator.ProduceExtractionError(
-          'unable to decode pair name with error: {0!s}'.format(exception))
-      return None, None
-
-    # Value: can be integer, boolean or text select by Type ID.
-    if type_id in (
-        self.TYPE_GENERAL_INTEGER, self.TYPE_INTEGER, self.TYPE_ENUMERATION):
-      value_structure = self.INTEGER
-    elif type_id == self.TYPE_BOOL:
-      value_structure = self.BOOLEAN
-    elif type_id == self.TYPE_DATETIME:
-      value_structure = self.DATETIME
-    else:
-      value_structure = self.TEXT
+    value = None
+    if attribute.tag_value in (0x21, 0x23):
```

Comment:
Please extract these tag value constants out, this is very hard to follow as is.